File: app/src/main/java/com/example/exitpro/Activity/HomeActivity.java
package com.example.exitpro.Activity;

import androidx.activity.result.ActivityResultLauncher;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import android.app.Dialog;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.Handler;
import android.util.Log;
import android.view.View;
import android.view.Window;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RelativeLayout;
import android.widget.Toast;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;
import com.example.exitpro.Utils.CaptureActUtil;
import com.example.exitpro.Config.Config;
import com.example.exitpro.Utils.FingerprintAuthHelperUtil;
import com.example.exitpro.GlobalVariables;
import com.example.exitpro.R;
import com.journeyapps.barcodescanner.ScanContract;
import com.journeyapps.barcodescanner.ScanOptions;

import org.json.JSONException;
import org.json.JSONObject;

public class HomeActivity extends AppCompatActivity {

    // UI elements
    Button btnOut, btnIn, btnLate, btnLogOut;
    RelativeLayout hHomeLayout;

    // Variables
    int scanNumber = -1;
    String destination = "";
    GlobalVariables globalVariables = new GlobalVariables();
    private ProgressDialog progressDialog;
    FingerprintAuthHelperUtil fingerprintAuthHelperUtil;

    // URLs for API requests
    public static String outURL = Config.BASE_URL + "/student/gate/exit";
    public static String inURL = Config.BASE_URL + "/student/gate/entry/";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_home);

        // Initialize UI elements
        btnOut = findViewById(R.id.btnOut);
        btnIn = findViewById(R.id.btnIn);
        btnLate = findViewById(R.id.btnLate);
        btnLogOut = findViewById(R.id.btnlogOut);
        hHomeLayout = findViewById(R.id.homeLayout);

        // Initialize fingerprint authentication
        fingerprintAuthHelperUtil = new FingerprintAuthHelperUtil(this, hHomeLayout);
        fingerprintAuthHelperUtil.authenticate();

        // Set up button listeners
        setupButtonListeners();

        // Check if the user is logged in
        if (!isLoggedIn()) {
            redirectToLoginActivity();
        }
    }

    @Override
    public void onRestart() {
        super.onRestart();
        fingerprintAuthHelperUtil.authenticate();
    }

    private void setupButtonListeners() {
        // Logout button listener
        btnLogOut.setOnClickListener(view -> logout());

        // Scan out button listener
        btnOut.setOnClickListener(view -> {
            scanNumber = -1;
            destination = "";
            startScan(outScan);
        });

        // Scan in button listener
        btnIn.setOnClickListener(view -> {
            scanNumber = -1;
            startScan(inScan);
        });

        // Latecomers button listener
        btnLate.setOnClickListener(view -> {
            Intent intent = new Intent(HomeActivity.this, LateComersActivity.class);
            startActivity(intent);
        });
    }

    private void startScan(ActivityResultLauncher<ScanOptions> scanLauncher) {
        ScanOptions options = new ScanOptions();
        options.setOrientationLocked(false);
        options.setPrompt("Scan a barcode");
        options.setCameraId(0); // Use a specific camera of the device
        options.setBeepEnabled(false);
        options.setBarcodeImageEnabled(true);
        options.setCaptureActivity(CaptureActUtil.class);
        scanLauncher.launch(options);
    }

    private final ActivityResultLauncher<ScanOptions> outScan = registerForActivityResult(new ScanContract(),
            result -> {
                if (result.getContents() != null) {
                    scanNumber = Integer.parseInt(result.getContents());
                    showDestinationDialog(String.valueOf(scanNumber));
                }
            });

    private final ActivityResultLauncher<ScanOptions> inScan = registerForActivityResult(new ScanContract(),
            result -> {
                if (result.getContents() != null) {
                    scanNumber = Integer.parseInt(result.getContents());
                    handleInScan();
                }
            });

    private void handleInScan() {
        showLoadingDialog();
        JSONObject jsonRequest = new JSONObject();
        RequestQueue queue = Volley.newRequestQueue(HomeActivity.this);

        // Create and send the JSON request for in scan
        JsonObjectRequest jsonObjectRequest = new JsonObjectRequest(
                Request.Method.PUT,
                inURL + scanNumber,
                jsonRequest,
                response -> {
                    dismissLoadingDialog();
                    handleInScanResponse(response);
                },
                error -> {
                    dismissLoadingDialog();
                    Toast.makeText(getApplicationContext(), "ERROR", Toast.LENGTH_SHORT).show();
                });

        queue.add(jsonObjectRequest);
    }

    private void handleInScanResponse(JSONObject response) {
        try {
            boolean success = response.getBoolean("isSuccess");
            if (success) {
                showSuccessDialog();
            } else {
                Toast.makeText(getApplicationContext(), "STUDENT IS INSIDE CAMPUS!", Toast.LENGTH_SHORT).show();
            }
        } catch (JSONException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        moveTaskToBack(true); // Move the task containing this activity to the back of the activity stack
    }

    private void showLoadingDialog() {
        progressDialog = new ProgressDialog(HomeActivity.this);
        progressDialog.setCancelable(false);
        progressDialog.setMessage("Please wait...");
        progressDialog.show();
    }

    private void dismissLoadingDialog() {
        if (progressDialog != null && progressDialog.isShowing()) {
            progressDialog.dismiss();
        }
    }

    private boolean isLoggedIn() {
        SharedPreferences sharedPreferences = getSharedPreferences("MyPrefs", Context.MODE_PRIVATE);
        return sharedPreferences.contains("access_token");
    }

    private void logout() {
        SharedPreferences sharedPreferences = getSharedPreferences("MyPrefs", Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.remove("access_token");
        editor.apply();
        redirectToLoginActivity();
    }

    private void redirectToLoginActivity() {
        Intent intent = new Intent(HomeActivity.this, LoginActivity.class);
        startActivity(intent);
        finish();
    }

    private void showDestinationDialog(final String scannedBarcode) {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle("Enter Destination");

        final EditText destinationInput = new EditText(this);
        builder.setView(destinationInput);

        builder.setPositiveButton("OK", (dialog, which) -> {
            destination = destinationInput.getText().toString();
            if (!destination.isEmpty()) {
                sendOutScanRequest();
            } else {
                Toast.makeText(getApplicationContext(), "DESTINATION IS INVALID", Toast.LENGTH_SHORT).show();
            }
        });

        builder.setNegativeButton("Cancel", (dialog, which) -> dialog.dismiss());

        builder.show();
    }

    private void sendOutScanRequest() {
        showLoadingDialog();
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put("roll_number", scanNumber);
            jsonObject.put("goingTo", destination);
        } catch (JSONException e) {
            throw new RuntimeException(e);
        }

        RequestQueue queue = Volley.newRequestQueue(HomeActivity.this);
        JsonObjectRequest jsonRequest = new JsonObjectRequest(Request.Method.POST, outURL, jsonObject,
                response -> {
                    dismissLoadingDialog();
                    handleOutScanResponse(response);
                },
                error -> {
                    dismissLoadingDialog();
                    Toast.makeText(getApplicationContext(), "ERROR", Toast.LENGTH_SHORT).show();
                });

        queue.add(jsonRequest);
    }

    private void handleOutScanResponse(JSONObject response) {
        try {
            boolean success = response.getBoolean("isSuccess");
            if (success) {
                showSuccessDialog();
            } else {
                Toast.makeText(getApplicationContext(), "STUDENT ALREADY OUT!", Toast.LENGTH_SHORT).show();
            }
        } catch (JSONException e) {
            throw new RuntimeException(e);
        }
    }

    private void showSuccessDialog() {
        Dialog dialog = new Dialog(this);
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
        dialog.setContentView(R.layout.success_dialog);
        dialog.setCancelable(false);
        dialog.getWindow().setBackgroundDrawableResource(android.R.color.transparent);
        dialog.show();

        new Handler().postDelayed(() -> dialog.dismiss(), 2000);
    }
}

File: app/src/main/java/com/example/exitpro/activity/HomeActivity.kt
package com.example.exitpro.activity

import android.app.Dialog
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.Window
import android.widget.Button
import android.widget.EditText
import android.widget.ProgressBar
import android.widget.RelativeLayout
import android.widget.TextView
import android.widget.Toast
import androidx.activity.OnBackPressedCallback
import androidx.activity.result.ActivityResultLauncher
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.lifecycle.Lifecycle
import com.example.exitpro.GlobalVariables
import com.example.exitpro.R
import com.example.exitpro.data.api.RetrofitClient
import com.example.exitpro.data.repository.ExitProRepository
import com.example.exitpro.utils.CaptureActUtil
import com.example.exitpro.utils.FingerprintAuthHelperUtil
import com.example.exitpro.utils.PermissionUtil
import com.example.exitpro.viewmodel.HomeViewModel
import com.example.exitpro.viewmodel.ViewModelFactory
import com.example.exitpro.viewmodel.state.ScanUiState
import com.journeyapps.barcodescanner.ScanContract
import com.journeyapps.barcodescanner.ScanIntentResult
import com.journeyapps.barcodescanner.ScanOptions
import kotlinx.coroutines.launch

@RequiresApi(Build.VERSION_CODES.P)
class HomeActivity : AppCompatActivity() {
    // UI elements
    private lateinit var btnOut: Button
    private lateinit var btnIn: Button
    private lateinit var btnLate: Button
    private lateinit var btnLogOut: Button
    private lateinit var guardNameView: TextView
    private lateinit var hHomeLayout: RelativeLayout

    // Variables
    private var scanNumber: Int = -1
    private var destination: String = ""
    private lateinit var globalVariables: GlobalVariables
    private var progressBar: ProgressBar? = null
    private var loadingDialog: Dialog? = null
    private lateinit var fingerprintAuthHelperUtil: FingerprintAuthHelperUtil
    
    // MVVM - ViewModel
    private val viewModel: HomeViewModel by lazy {
        val repository = ExitProRepository(RetrofitClient.apiService)
        val factory = ViewModelFactory(repository)
        factory.create(HomeViewModel::class.java)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_home)

        // Initialize UI elements
        btnOut = findViewById(R.id.btnOut)
        btnIn = findViewById(R.id.btnIn)
        btnLate = findViewById(R.id.btnLate)
        btnLogOut = findViewById(R.id.btnlogOut)
        hHomeLayout = findViewById(R.id.homeLayout)
        guardNameView = findViewById(R.id.guard_name_view)

        // Initialize global variables
        globalVariables = GlobalVariables()

        // Initialize fingerprint authentication helper
        fingerprintAuthHelperUtil = FingerprintAuthHelperUtil(this, hHomeLayout)
        fingerprintAuthHelperUtil.authenticate()

        // Set up button listeners
        setupButtonListeners()

        // Check if the user is logged in
        if (!isLoggedIn) {
            redirectToLoginActivity()
        }
        
        // Configure back button behavior
        setupBackPressHandler()
        
        // Request necessary runtime permissions
        requestRuntimePermissions()
        
        // Observe ViewModel states
        observeViewModelStates()
    }

    override fun onRestart() {
        super.onRestart()
        fingerprintAuthHelperUtil.authenticate()
    }

    /**
     * Set up listeners for the buttons.
     */
    private fun setupButtonListeners() {
        // Logout button listener
        btnLogOut.setOnClickListener { logout() }

        // Scan out button listener
        btnOut.setOnClickListener {
            scanNumber = -1
            destination = ""
            startScan(outScan)
        }

        // Scan in button listener
        btnIn.setOnClickListener {
            scanNumber = -1
            startScan(inScan)
        }

        // Latecomers button listener
        btnLate.setOnClickListener {
            val intent = Intent(this, LateComersActivity::class.java)
            startActivity(intent)
        }
    }

    /**
     * Start the barcode scanning process.
     *
     * @param scanLauncher The launcher for the scan activity.
     */
    private fun startScan(scanLauncher: ActivityResultLauncher<ScanOptions>) {
        val options = ScanOptions().apply {
            setOrientationLocked(false)
            setPrompt("Scan a barcode")
            setCameraId(0) // Use a specific camera of the device
            setBeepEnabled(false)
            setBarcodeImageEnabled(true)
            setCaptureActivity(CaptureActUtil::class.java)
        }
        scanLauncher.launch(options)
    }

    private val outScan = registerForActivityResult(
        ScanContract()
    ) { result: ScanIntentResult ->
        result.contents?.let {
            scanNumber = it.toInt()
            showDestinationDialog(scanNumber.toString())
        }
    }

    private val inScan = registerForActivityResult(
        ScanContract()
    ) { result: ScanIntentResult ->
        result.contents?.let {
            scanNumber = it.toInt()
            handleInScan()
        }
    }

    /**
     * Handle the scan-in process using ViewModel (MVVM pattern).
     */
    private fun handleInScan() {
        viewModel.processStudentEntry(scanNumber)
    }

    /**
     * Request runtime permissions required by the app.
     * Includes camera for scanning and notifications.
     */
    private fun requestRuntimePermissions() {
        // Request camera permission if not granted
        if (!PermissionUtil.checkCameraPermission(this)) {
            PermissionUtil.requestCameraPermission(this)
        }
        
        // Request notification permission on Android 13+
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (!PermissionUtil.checkNotificationPermission(this)) {
                PermissionUtil.requestNotificationPermission(this)
            }
        }
    }
    
    /**
     * Configure back button behavior to minimize app instead of closing.
     */
    private fun setupBackPressHandler() {
        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                moveTaskToBack(true)
            }
        })
    }

    /**
     * Display a loading dialog while processing.
     */
    private fun showLoadingDialog() {
        loadingDialog = Dialog(this).apply {
            requestWindowFeature(Window.FEATURE_NO_TITLE)
            setContentView(R.layout.loading_dialog)
            setCancelable(false)
            window?.setBackgroundDrawableResource(android.R.color.transparent)
            show()
        }
    }

    /**
     * Dismiss the loading dialog if currently showing.
     */
    private fun dismissLoadingDialog() {
        loadingDialog?.takeIf { it.isShowing }?.dismiss()
    }

    /**
     * Check if the user is logged in.
     *
     * @return True if the user is logged in, false otherwise.
     */
    private val isLoggedIn: Boolean
        get() {
            val sharedPreferences = getSharedPreferences("MyPrefs", MODE_PRIVATE)
            if (sharedPreferences.contains("guard_name")) {
                guardNameView.text = "Welcome, ${sharedPreferences.getString("guard_name", "")}!"
                return true
            }
            return false
        }

    /**
     * Log out the user and redirect to the login activity.
     */
    private fun logout() {
        val sharedPreferences = getSharedPreferences("MyPrefs", MODE_PRIVATE)
        with(sharedPreferences.edit()) {
            remove("otp")
            remove("guard_name")
            apply()
        }
        redirectToLoginActivity()
    }

    /**
     * Redirect to the login activity.
     */
    private fun redirectToLoginActivity() {
        startActivity(Intent(this, LoginActivity::class.java))
        finish()
    }

    /**
     * Show a dialog to enter the destination after scanning a barcode.
     *
     * @param scannedBarcode The scanned barcode.
     */
    private fun showDestinationDialog(scannedBarcode: String) {
        val builder = AlertDialog.Builder(this).apply {
            setTitle("Enter Destination")
            val destinationInput = EditText(context)
            setView(destinationInput)

            setPositiveButton("OK") { dialog, _ ->
                destination = destinationInput.text.toString()
                if (destination.isNotEmpty()) {
                    sendOutScanRequest()
                } else {
                    Toast.makeText(applicationContext, "DESTINATION IS INVALID", Toast.LENGTH_SHORT).show()
                }
            }

            setNegativeButton("Cancel") { dialog, _ -> dialog.dismiss() }
        }

        builder.show()
    }

    /**
     * Send the out-scan request to the server using ViewModel (MVVM pattern).
     */
    private fun sendOutScanRequest() {
        viewModel.processStudentExit(scanNumber, destination)
    }

    /**
     * Show a success dialog for a short duration.
     */
    private fun showSuccessDialog() {
        val dialog = Dialog(this).apply {
            requestWindowFeature(Window.FEATURE_NO_TITLE)
            setContentView(R.layout.success_dialog)
            setCancelable(false)
            window?.setBackgroundDrawableResource(android.R.color.transparent)
            show()
        }

        Handler(Looper.getMainLooper()).postDelayed({ dialog.dismiss() }, 2000)
    }

    /**
     * Observe ViewModel state changes and update UI accordingly (MVVM pattern).
     */
    private fun observeViewModelStates() {
        // Observe student entry state
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.entryState.collect { state ->
                    when (state) {
                        is ScanUiState.Idle -> {
                            // Do nothing
                        }
                        is ScanUiState.Loading -> {
                            showLoadingDialog()
                        }
                        is ScanUiState.Success -> {
                            dismissLoadingDialog()
                            showSuccessDialog()
                            viewModel.resetEntryState()
                        }
                        is ScanUiState.Error -> {
                            dismissLoadingDialog()
                            val message = if (state.message == "Student is inside campus") {
                                "STUDENT IS INSIDE CAMPUS!"
                            } else {
                                state.message
                            }
                            Toast.makeText(applicationContext, message, Toast.LENGTH_SHORT).show()
                            viewModel.resetEntryState()
                        }
                    }
                }
            }
        }

        // Observe student exit state
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.exitState.collect { state ->
                    when (state) {
                        is ScanUiState.Idle -> {
                            // Do nothing
                        }
                        is ScanUiState.Loading -> {
                            showLoadingDialog()
                        }
                        is ScanUiState.Success -> {
                            dismissLoadingDialog()
                            showSuccessDialog()
                            viewModel.resetExitState()
                        }
                        is ScanUiState.Error -> {
                            dismissLoadingDialog()
                            Toast.makeText(applicationContext, state.message, Toast.LENGTH_SHORT).show()
                            viewModel.resetExitState()
                        }
                    }
                }
            }
        }
    }

    /**
     * Handle permission request results from the user.
     */
    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        
        when (requestCode) {
            PermissionUtil.REQUEST_CAMERA_PERMISSION -> {
                PermissionUtil.onRequestPermissionsResult(
                    requestCode, permissions, grantResults,
                    onGranted = {
                        Toast.makeText(this, "Camera permission granted", Toast.LENGTH_SHORT).show()
                    },
                    onDenied = {
                        Toast.makeText(this, "Camera permission is required for scanning QR codes", Toast.LENGTH_LONG).show()
                    }
                )
            }
            PermissionUtil.REQUEST_NOTIFICATION_PERMISSION -> {
                PermissionUtil.onRequestPermissionsResult(
                    requestCode, permissions, grantResults,
                    onGranted = {
                        // Notification permission granted
                    },
                    onDenied = {
                        Toast.makeText(this, "Notification permission denied. You may not receive important updates.", Toast.LENGTH_LONG).show()
                    }
                )
            }
        }
    }

}